flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/RowTimeBoundedStreamInnerJoin.scala

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.join
+
+import org.apache.flink.api.common.typeinfo.TypeInformation
+import org.apache.flink.streaming.api.functions.co.CoProcessFunction
+import org.apache.flink.table.runtime.types.CRow
+import org.apache.flink.types.Row
+
+/**
+  * The function to execute row(event) time bounded stream inner-join.
+  */
+class RowTimeBoundedStreamInnerJoin(
+    leftLowerBound: Long,
+    leftUpperBound: Long,
+    allowedLateness: Long,
+    leftType: TypeInformation[Row],
+    rightType: TypeInformation[Row],
+    genJoinFuncName: String,
+    genJoinFuncCode: String,
+    leftTimeIdx: Int,
+    rightTimeIdx: Int)
+    extends TimeBoundedStreamInnerJoin(
+      leftLowerBound,
+      leftUpperBound,
+      allowedLateness,
+      leftType,
+      rightType,
+      genJoinFuncName,
+      genJoinFuncCode,
+      leftTimeIdx,
+      rightTimeIdx,
+      JoinTimeIndicator.ROWTIME) {
+
+  override def checkRowOutOfDate(timeForRow: Long, watermark: Long) = {
+    timeForRow <= watermark - allowedLateness
+  }
+
+  override def updateOperatorTime(ctx: CoProcessFunction[CRow, CRow, CRow]#Context): Unit = {
+    rightOperatorTime =
+      if (ctx.timerService().currentWatermark() > 0) ctx.timerService().currentWatermark()
```

Comment:
Yes, you are right. I'll move this check to places where we actually use the watermark.

Reply:
After checking the codes, I find out that the number of watermark usages is even larger than that of their updates (i.e., each update is followed by at least one usage). Considering that the number of usage may increase in the future, I suggest to keep this check in the update method. Of course, it would be better if we can use a "safer" initial value for
 the watermark😄. What do you think?

Reply:
OK, let's keep it here. Changing the value of watermarks won't be possible as it is built into the DataStream API and some users rely on the current behavior. The curse of public APIs ;-)

Reply:
Totally understand 😄